### PR TITLE
Update testing to use Catch2 and CMake

### PR DESCRIPTION
**Pull request by AidanBeltonS**
This PR moves the testing away from using `make` and utility functions. It introduces Catch2 as a git submodule so the user does not have to separately build it. CMake is then used to build and run the tests. Using Catch2 reduces the amount of code needed to test `SyclCPLX` and will speed up adding new tests.

The readme has been updated to reflect this change.

### CODE REVIEW

**Comment by TApplencourt** [comment]
Back in vacation, will have a look :) Thanks!

**Comment by TApplencourt** [comment]
Should we use `fetchContent`? (https://github.com/catchorg/Catch2/blob/devel/docs/cmake-integration.md) I know nothing about CMAKE, but I know that I hate git submodule (always forgot to put the --recursive)

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,45 @@
+file(GLOB test_cases CONFIGURE_DEPENDS "*.cpp")
+
+# set(test_cases
```
Comment (TApplencourt) [inline tests/CMakeLists.txt:3]: I guess we can remove the comment now :)

File: tests/test_helper.hpp [hunk 2]
```diff
@@ -6,12 +9,42 @@
...
 
-template <typename T> const char *get_typename() { return "Unknown type"; }
-template <> const char *get_typename<double>() { return "double"; }
-template <> const char *get_typename<float>() { return "float"; }
-template <> const char *get_typename<sycl::half>() { return "sycl::half"; }
+template <> inline bool is_type_supported<double>(sycl::queue &Q) {
+  return Q.get_device().has(sycl::aspect::fp64);
+}
+
+template <> inline bool is_type_supported<float>(sycl::queue &Q) {
```
Comment (TApplencourt) [inline tests/test_helper.hpp:19]: I never realized that indeed fp32 is required by the SYCL spec :D

File: tests/test_complex_types.cpp [hunk 3]
```diff
@@ -2,120 +2,75 @@
...
-    static_assert(
-        std::is_same_v<complex<T>, decltype(pow(complex<T>(), complex<T>()))>);
-    static_assert(std::is_same_v<complex<T>, decltype(pow(T(), complex<T>()))>);
-    return true;
-  }
-};
-
-// Check functions return correct types
-void check_math_function_types() {
+TEST_MATH_FUNC_TYPE("Test complex acos types", "[acos]", acos)
```
Comment (TApplencourt) [inline tests/test_complex_types.cpp:35]: Maybe a stupid question, but it seems strange to me that if we list all the `TEST_MATH_FUNC_TYPE` we need to add them manually in the CMAKELIST. 
Sorry, I'm sure I'm missing something obvious...

File: tests/test_helper.hpp [hunk 4]
```diff
@@ -6,12 +9,42 @@
...
+template <> inline bool is_type_supported<sycl::half>(sycl::queue &Q) {
+  return Q.get_device().has(sycl::aspect::fp16);
+}
+
+// Helper for passing infinity and nan values
+template <typename T>
+inline constexpr T inf_val = std::numeric_limits<T>::infinity();
+
+template <typename T>
+inline constexpr T nan_val = std::numeric_limits<T>::quiet_NaN();
```
Comment (TApplencourt) [inline tests/test_helper.hpp:32]: Oh nice, I did new about `::quiet_Nan()`. I was using `std::nan("")` like a caveman :)

File: tests/acosh_complex.cpp [hunk 5]
```diff
@@ -1,21 +1,37 @@
...
+       make_tuple(cmplx<T>{nan_val<T>, nan_val<T>}, true),
+       make_tuple(cmplx<T>{nan_val<T>, inf_val<T>}, true),
+       make_tuple(cmplx<T>{inf_val<T>, nan_val<T>}, true)}));
+
+  auto std_in = init_std_complex(input);
+  sycl::ext::cplx::complex<T> cplx_input{input.re, input.im};
 
-    std::complex<T> std_out{init_re, init_im};
-    auto *cplx_out = sycl::malloc_shared<sycl::ext::cplx::complex<T>>(1, Q);
+  std::complex<T> std_out{input.re, input.im};
```
Comment (TApplencourt) [inline tests/acosh_complex.cpp:26]: I guess we can just do `std_out{input}`

File: tests/acosh_complex.cpp [hunk 6]
```diff
@@ -26,47 +42,18 @@ template <typename T> struct test_acosh {
       });
     }
     Q.wait();
```
Comment (TApplencourt) [inline tests/acosh_complex.cpp:44]: Just for elegance for `Q.wait()` should be inside the `if(is_error_checking)`, but not a big deal

**Comment by TApplencourt** [comment]
Thanks for all ctest / catch integration. Look far better now :)

File: tests/CMakeLists.txt (hunk 1 above)
Comment (AidanBeltonS) [reply tests/CMakeLists.txt]: Haha thanks for catching this. I have removed the comments

File: tests/test_complex_types.cpp (hunk 3 above)
Comment (AidanBeltonS) [reply tests/test_complex_types.cpp:35]: So I don't quite understand. `TEST_MATH_FUNC_TYPE` is a macro that just allows the test to be repeated for all the different functions without me re-writing them. CMake needs a list of all the source files which then Catch2 parses for the tests Catch2 looks for a different macro something along the line of `TEST_CASE("Test is_gencomplex", "[gencomplex]")`.
CMake needs a list of the test source files so it knows which files to compile otherwise it is will compile none of the tests files.
Does this answer your question, could you clarify if it does not?

File: tests/acosh_complex.cpp (hunk 5 above)
Comment (AidanBeltonS) [reply tests/acosh_complex.cpp:26]: So unfortunately I cant do `std_out{input}` as input is of type `cmplx` which is just a helper struct of real and imaginary components. std::complex does not have a constructor for my helper struct `cmplx`. 

Also in this case we don't want the `half's` to be be translated into `floats` as this is the output value which should have the half type. As no operation is being performed on the output type this is valid, the reason we are initializing it to the input values is for the case were we are checking if `acosh` can undo `cosh`. However when error codes are tested this value is overwritten.

File: tests/acosh_complex.cpp [hunk 7]
```diff
@@ -1,21 +1,37 @@
...
-    std::complex<T> std_out{init_re, init_im};
-    auto *cplx_out = sycl::malloc_shared<sycl::ext::cplx::complex<T>>(1, Q);
+  std::complex<T> std_out{input.re, input.im};
+  auto *cplx_out = sycl::malloc_shared<sycl::ext::cplx::complex<T>>(1, Q);
 
-    // Get std::complex output
-    if (is_error_checking)
-      std_out = std::acosh(std_in);
+  // Get std::complex output
+  if (is_error_checking)
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply tests/acosh_complex.cpp:30]: I don't think `is_type_supported` should protect this, as lower down we test if the behaviour is valid on the host. While SYCL does say that a device does not need to support `double` and `half` types the host does so it makes sense not to calculate the reference value and have it for the later host check. It simply saves having two places where the reference value is calculated.

File: README.md [hunk 8]
```diff
@@ -44,4 +44,17 @@ int main() {
...
+```
+mkdir build
+
+cd build
+
+cmake -G Ninja -DCMAKE_CXX_COMPILER=$PATH_TO_LLVM/bin/clang++ -DCMAKE_CXX_FLAGS=-fsycl ..
+
+ninja
+
+ctest
```
Comment (TApplencourt) [inline README.md:59]: Do you mind using `make` for the doc? HPC user are old :) 
Also, removing the `clang++` as it's named `icpx` or `dpcpp` for me

```suggestion
mkdir build
cd build
cmake -DCMAKE_CXX_COMPILER=$CXX_PATH -DCMAKE_CXX_FLAGS=-fsycl ..
make -j 8
ctest
```

File: tests/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    add_executable(${exe_name} ${test_file})
+    target_include_directories(${exe_name} PUBLIC ../include/)
+    target_link_libraries(${exe_name} PRIVATE
+        Catch2::Catch2WithMain
+    )
+    
+    catch_discover_tests(${exe_name})
+    # else()
+        # message(FATAL_ERROR "No file named ${test_file}")
+    # endif()
```
Comment (TApplencourt) [inline tests/CMakeLists.txt:16]:
```suggestion
```

**Comment by TApplencourt** [comment]
We saw `32% tests passed, 179 tests failed out of 262` with `icpx` / `dpcpp`. Look like oneAPI is not happy with CATCH2... Investigating right now. (all the old tests passed)

File: tests/test_complex_types.cpp (hunk 3 above)
Comment (TApplencourt) [reply tests/test_complex_types.cpp:35]: Oh I see. Yes it clarifies; thanks! `TEST_MATH_FUNC_TYPE(acos)` is not calling or registering the tests inside `acos_complex.cpp`. 

TEST_MATH_FUNC_TYPE are just another type of tests. 

Thanks!

**Comment by TApplencourt** [comment]
Filled https://github.com/intel/llvm/issues/6720

File: tests/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,17 @@
+file(GLOB test_cases CONFIGURE_DEPENDS "*.cpp")
+
+foreach(test_file IN LISTS test_cases)
+    # if(EXISTS "${test_file}")
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply tests/CMakeLists.txt]: I have uncommented these checks to make sure each file is available and emit a warning if it is not.

File: tests/CMakeLists.txt (hunk 9 above)
Comment (AidanBeltonS) [reply tests/CMakeLists.txt]: I have uncommented this check. Same as above.

**Comment by AidanBeltonS** [comment]
Thanks for filing the ticket. I do not have these issues locally, I suppose the issue is specific to the distributed version of `dpcpp`.

**Comment by TApplencourt** [comment]
>Thanks for filing the ticket. I do not have these issues locally. I am building dpcpp from the open source repo, so I guess the issue is specific to the distributed version of dpcpp.

Did you use the L0 backend? It's maybe backend specific. But We have a work-around ( just run the test one by one), so I will write a little script to do so, so I can continue testing with `dpcpp`. 

Sorry for the late merge! And thanks again

**Comment by AidanBeltonS** [comment]
> > Thanks for filing the ticket. I do not have these issues locally. I am building dpcpp from the open source repo, so I guess the issue is specific to the distributed version of dpcpp.
> 
> Did you use the L0 backend? It's maybe backend specific. But We have a work-around ( just run the test one by one), so I will write a little script to do so, so I can continue testing with `dpcpp`. EDIT: I Just tested dpcpp with OpenCL backend, and it still segfault. So I Agree with you, looks like a packaging issue.
> 
> Sorry for the late merge! And thanks again

Yes, I have been testing with the L0 backend on a local machine
